src/ddsrt/CMakeLists.txt

```diff
@@ -9,132 +9,114 @@
 # runtime must be instructed to use it instead of the native stack. Of course
 # for embedded targets there is no "native" stack and the runtime module must
 # always be instructed to use an "alternative" stack.
 option(WITH_LWIP "Use lightweight IP stack" OFF)
-option(WITH_DNS "Enable domain name lookups" ON)
 option(WITH_FREERTOS "Build for FreeRTOS" OFF)
 
-function(check_runtime_feature SOURCE_FILE)
-  get_target_property(_defs ddsrt INTERFACE_COMPILE_DEFINITIONS)
-  foreach(_def ${_defs})
-    set(_strdefs "${_strdefs} -D${_def}")
-  endforeach()
-
-  # Generate dummy export header required by feature tests.
-  generate_dummy_export_header(
-    ddsrt
-    BASE_NAME dds
-    EXPORT_FILE_NAME "${CMAKE_CURRENT_BINARY_DIR}/cmake/include/dds/export.h")
-  set(_strincs "${CMAKE_CURRENT_BINARY_DIR}/cmake/include")
-
-  get_target_property(_incs ddsrt INTERFACE_INCLUDE_DIRECTORIES)
-  foreach(_inc ${_incs})
-    set(_strincs "${_strincs};${_inc}")
-  endforeach()
-
-  if(_strincs)
-    set(_strincs "-DINCLUDE_DIRECTORIES:STRING=${_strincs}")
-  endif()
-
-  set(expr "cmake_([_a-zA-Z0-9]+)=([_a-zA-Z0-9]+)")
-  try_compile(
-    foo "${CMAKE_BINARY_DIR}"
-    SOURCES "${CMAKE_CURRENT_SOURCE_DIR}/${SOURCE_FILE}"
-    CMAKE_FLAGS "${_strincs}"
-    COMPILE_DEFINITIONS "${_strdefs}"
-    OUTPUT_VARIABLE output)
-  string(REGEX MATCHALL "${expr}" matches "${output}")
-  foreach(match ${matches})
-    string(REGEX REPLACE "${expr}" "\\1" option "${match}")
-    string(REGEX REPLACE "${expr}" "\\2" value "${match}")
-    set(${option} ${value} PARENT_SCOPE)
-  endforeach()
-endfunction()
-
-if(WITH_FREERTOS)
-  set(system_name freertos)
-elseif(APPLE)
-  set(system_name darwin)
-elseif(ANDROID)
-  # FIXME: Not correct, but will do for the short-term. A better way would be
-  #        fallback to linux, and then posix.
-  set(system_name linux)
+if(WITH_LWIP)
+  set(event_backend "select")
+elseif(CMAKE_SYSTEM MATCHES "Linux" OR
+       CMAKE_SYSTEM MATCHES "Windows")
+  set(event_backend "epoll")
+  set(DDSRT_HAVE_NETLINK_EVENT TRUE)
+elseif(CMAKE_SYSTEM MATCHES "Darwin" OR
+       CMAKE_SYSTEM MATCHES "BSD")
+  set(event_backend "kqueue")
+  set(DDSRT_HAVE_NETLINK_EVENT TRUE)
 else()
-  string(TOLOWER ${CMAKE_SYSTEM_NAME} system_name)
+  set(event_backend "select")
 endif()
 
-# A better choice is to use a so-called object library for ddsrt, but it was
-# not possible to use target_link_libraries with object libraries until CMake
-# 3.12. At the time of this writing most long-term stable distributions still
-# ship an older version, so an interface library with public sources is used
-# as a workaround for now.
-add_library(ddsrt INTERFACE)
+set(WITH_EVENT "${event_backend}" CACHE STRING "Use specified backend for event framework")
+set_property(CACHE WITH_EVENT PROPERTY STRINGS "epoll" "kqueue" "select")
+# discovery + data + pipe + netlink == 4
+set(FIXED_EVENTS "4" CACHE STRING "Minimum number of events to reserve space for")
```

Comment:
I haven't checked what constitutes an event, but it is not unusual to have 4 sockets, which would suggest it makes more sense to reserve 6 rather than 4.